docs/source/notes/extending.rst

```diff
@@ -12,53 +12,97 @@ Extending :mod:`torch.autograd`
-  as many arguments as you want, with some of them being optional, if you
-  specify the default values. All kinds of Python objects are accepted here.
-  :class:`Tensor` arguments that track history (i.e., with
-  ``requires_grad=True``) will be converted to ones that don't track history
-  before the call, and their use will be registered in the graph. Note that this
-  logic won't traverse lists/dicts/any other data structures and will only
-  consider :class:`Tensor` s that are direct arguments to the call. You can
-  return either a single :class:`Tensor` output, or a :class:`tuple` of
-  :class:`Tensor` s if there are multiple outputs. Also, please refer to the
-  docs of :class:`Function` to find descriptions of useful methods that can be
-  called only from :meth:`~Function.forward`.
-- :meth:`~Function.backward` - gradient formula. It will be given
-  as many :class:`Tensor` arguments as there were outputs, with each of them
-  representing gradient w.r.t. that output. It should return as many
-  :class:`Tensor` s as there were inputs, with each of them containing the
-  gradient w.r.t. its corresponding input. If your inputs didn't require
-  gradient (:attr:`~ctx.needs_input_grad` is a tuple of booleans indicating
-  whether each input needs gradient computation), or were non-:class:`Tensor`
-  objects, you can return :class:`python:None`. Also, if you have optional
-  arguments to :meth:`~Function.forward` you can return more gradients than there
-  were inputs, as long as they're all :any:`python:None`.
-
-.. note::
-
-  It's the user's responsibility to use the special functions in the forward's `ctx`
-  properly in order to ensure that the new :class:`Function` works properly with
-  the autograd engine.
-
-  - :meth:`~torch.autograd.function._ContextMethodMixin.save_for_backward` must be
-    used when saving input or output of the forward to be used later in the backward.
-  - :meth:`~torch.autograd.function._ContextMethodMixin.mark_dirty` must be used to
-    mark any input that is modified inplace by the forward function.
-  - :meth:`~torch.autograd.function._ContextMethodMixin.mark_non_differentiable` must
-    be used to tell the engine if an output is not differentiable.
-  - :meth:`~torch.autograd.function._ContextMethodMixin.set_materialize_grads` can be
-    used to tell the autograd engine to optimize gradient computations in the cases where
-    the output does not depend on the input by not materializing grad tensors given to backward
-    function. That is, if set to False, None object in python or "undefined tensor" (tensor x for
-    which x.defined() is False) in C++ will not be converted to a tensor filled with zeros prior
-    to calling backward. However, supporting this optimization means your custom autograd function
-    has to handle gradients that are represented in this way and is thus opt-in. Default value is True.
+Subclassing :class:`autograd.Function` defines a custom autograd function.
```

Comment:
I think this intro sentence could be better. Especially because "custom autograd function" is not defined here.
The sentence explaining that Functions is what autograd uses to compute the result and gradients and encode the history sounds better?